### ISSUE TITLE
[Bug] Type definition reports spurious errors 2536 and 2322 but type works as expected in actual usage

### ISSUE THREAD

**Issue by marcopullo** [microsoft/TypeScript#49303]
# Bug Report

Type definition reports spurious errors 2536 and 2322 but the type works as expected in actual usage.

### 🔎 Search Terms

2536
2322

### 🕗 Version & Regression Information

Reproduced in:
- Version 4.7.2
- v4.8.0-dev.20220529

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220529#code/PTAEBUE8AcFMGcDGAnAltALqRB7AttKgDazKgDmqAbqgHbminI7Lyg62gYAWsoAgvACi-AGpcYfACawAZnVQZUHADSgARgFcsdUNGQBDREsR9FoAO4sA1mwNtYADzjHYUgFAhLNg8020pAC5QTXg+AAEMeFRyWhY+DBxQGLjkBN5GZGZWUF409wxJUAB5dQArWGNwHHA0aBIAHn5GRwxYALYAb1BaAzwEaCNYYPgMNHpQAF81ACFQAF5QUfGGADJQa1hIHFkBAD4F0DmnNo7QA1pIUAB+UABtWk08dVI1AAMAEk7+O4AiXv68EGpl+AF1JsAvjNJm81D8ZqDQaBgrRYFRSABudwFIqlMqHTruUDEnp9AZDYIAcngfXqsEpKiJJLarRGYzo5EZkxxcFAAGVaSRaugSIc8ZUMNVhXSGni9tivOKqjU6qKrMhbOcHM4JW5kpwDEQiNh7Ah3LhaKMWoLhvybdKSHckYs7kziV5RIbUFJQEJHDb4G77gBGNTUm3AAHk0wM0DhggkSmgtQk0BeQouXiITX2FouNoeVN3UNxmkJ2DAFkYWPxulcE5JlMk9OSRBZnPa-NubGprwASVoVC9PqcAaDxbDZbpkbJQKGNb7ADlRPwADJ9gAijebYAzfCYLC1ed1hZJE9LEarseDAA4AOwAZgATMnU2+dxJeQeyLmnF2PKC7hAA)

### 💻 Code

```ts
// Typescript compiler giving errors on the ObjectToTriple type definition, but in practice it works as expected
//     1. `Type 'B' is not assignable to type 'string | number | bigint | boolean | null | undefined'.(2322)`
//     2. `Type 'B' cannot be used to index type 'A'.(2536)`
// workaround: use @tsignore to ignore the errors here
type ObjectToTriple<A extends { namespace: string }, B = string & keyof A> = B extends any ? [number, `${A["namespace"]}/${B}`, A[B]] : never;

type Obj = {
    namespace: 'sample',
    text: string,
}

// ObjectToTriple works as expected in all cases
const example: ObjectToTriple<Obj>[] = [
    // Valid Examples
    [1, 'sample/namespace', 'sample'],     // typechecks as expected
    [1, 'sample/text', 'sample text'],     // typechecks as expected

    // Invalid examples
    [1, 'sample/namespace', 'INVALID'],   // type error as expected
    [1, 'sample/text', 18732],            // type error as expected
]

```

### 🙁 Actual behavior

Two Typescript compiler errors reported in the definition of `ObjectToTriple`, despite the type actually working in practice as expected. 
Workaround: `@tsignore` to ignore the type definition errors and use the type as expected.

### 🙂 Expected behavior

No Typescript errors in the definition of `ObjectToTriple`.

**Comment by fatcerberus** [comment]
Despite having a default, `B` is unconstrained and could be anything.  In particular it might include symbols, which can't be used in a template literal, or it might be some object type, which can't be used as an indexer.  In particular, `ObjectToTriple<Obj, Obj>` is a valid instantiation, in which case the type alias as written is nonsensical.  Hence the errors.

This makes all the errors go away:
```ts
type ObjectToTriple<A extends { namespace: string }, B extends string & keyof A = string & keyof A> = B extends any ? [number, `${A["namespace"]}/${B}`, A[B]] : never;
```

**Comment by marcopullo** [comment]
Thanks for the clarification @fatcerberus! That explanation helped confirm it was indeed my confusion over default instantiation but not actually constraining the type.
